### PR TITLE
rfc: structuring Tendermint as a library

### PR DESCRIPTION
**Pull request by cmwaters**
This RFC attempts to capture some of the previous discussions around how we want to mold Tendermint for its users. I would like people to add their own view and use the discussion to be more deliberate about how we structure the codebase.

### CODE REVIEW

**Comment by tychoish** [comment]
I think there are some interesting additional questions that we'd want to recon with: 
- the time to delivery on a stable, well designed service (given the smaller surface area) is going to be much shorter. 
-  there is some (potential) cost in the library model, of rendering APIs less stable (e.g. adding features requires some change to APIs, and if we don't account for then we either break things more often then user's like (and therefore slowing adoption), have implementations that are more convoluted to prevent breaking APIs, or are forced to deliver features based on tolerance for API change rather than potential impact.  Or more likely some combination of this. 
- the real consideration, I think, is what downstream users can tolerate and are willing to support. I don't neccessarily think that (go) API stability should be our leading goal, but I it's easy to make a service a library, and harder to reduce the surface area of an API.
- additionally if tendermint is a library and consists of composeble components that the cost of testing (and support) will increase, as the number of operational configurations will increase. This probably comes at the cost of some development velocity and perceived reliability.

There are a couple of existential questions here: first that about plug-ability, which I think Tess' document of a few years ago addresses pretty well, and then secondly, "how should we, as a project, feel about forks." Obviously long-term forks are a problem in some respect (as they represent conflicting views about the goals of a project), but smaller forks that maintain small patch-sets of changes to address specific needs are really easy to maintain, and perhaps for projects like tendermint (and the SDK!) we should encourage these second kind of fork as a way of encouraging users to experiment and expediently develop changes to tendermint. There are, perhaps, ways that we can find a path between these two options (forking and plugability) that resolve the larger question (service or library?) in a way that supports actual use cases.

**Review by marbar3778 — COMMENTED** [review]
Hmm I would disagree with the tradeoffs. I think having external interfaces for intermodule dependence and internal implementations maintained by the team is the ideal scenario. This provides the best of both worlds.

**Comment by cmwaters** [comment]
>  I think having external interfaces for intermodule dependence and internal implementations maintained by the team is the ideal scenario.

This is the path I assume we would go down as well. What part of the tradeoffs do you disagree with?

**Comment by tychoish** [comment]
While I think we can make different decisions and value jugements about the trade-offs and values provided by different designs, but I think the advantages and disadvantages are still there.

While it may be somewhat out of scope for this document, while it's possible to sort of decide on a goal, the path *toward* that goal is important for considering the practicality of that goal. Is this a "rip up the carpet and spend 6 months moving code around" or is this "iterate slowly over 24 months of quarterly releases" and having answers or methods for determining the answers to questions like "how much churn can users expect on the interfaces between components?" Make it easier to align on solutions or feasibility from either engineering or product perspectives.

I think we all agree on the goal, broadly speaking, but to make a decision on a project like this we also have to understand the engineering costs.

File: docs/rfc/rfc-022-structuring-as-a-library.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# RFC 022: Structuring Tendermint as a Library
+
+## Changelog
+
+- 25 July 2022: Initial draft (@cmwaters)
+
+## Abstract
+
+How do we want our users to use Tendermint? How does Tendemrint best deliver it's unique value proposition in a way that moves it towards being the leading generalizable consensus engine? This document aims to explore these themes mainly through comparison of Tendermint as a library or Tendermint as a service. The RFC focuses just on product direction and offers no concrete technical solutions although decisions here will dictate future technical discussions as we look to improve the ergonomics of the repository.
```
Comment (thanethomson) [inline docs/rfc/rfc-022-structuring-as-a-library.md:9]:
```suggestion
How do we want our users to use Tendermint? How does Tendermint best deliver it's unique value proposition in a way that moves it towards being the leading generalizable BFT consensus engine? This document aims to explore these themes mainly through comparison of Tendermint as a library or Tendermint as a service. The RFC focuses just on product direction and offers no concrete technical solutions although decisions here will dictate future technical discussions as we look to improve the ergonomics of the repository.
```

File: docs/rfc/rfc-022-structuring-as-a-library.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+## Background
+
+The word "library" is not used once in the codebases' documentation, however, it is understood that this was perhaps the intention of the original authors. As the codebase expanded and adoption grew, there became a need to distinguish between private and public APIs to offer some stability whilst still developing the software. This acculminated in [ADR 060](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-060-go-api-stability.md) which internalized the majority of functionality behind an `internal` directory. This reflected the thinking at the time which endeavoured to present Tendermint as a simple service that one could run without having to worry about what was really happening. For those that had business concerns that required more than what was offered, they had to fork the codebase as is evident with [Celestia](https://github.com/celestiaorg/celestia-core). While there has been numerous discussions over the past years, nothing has been explicitly documented as this RFC aims to do.
```
Comment (thanethomson) [inline docs/rfc/rfc-022-structuring-as-a-library.md:13]:
```suggestion
The word "library" is not used once in the codebases' documentation, however, it is understood that this was perhaps the intention of the original authors. As the codebase expanded and adoption grew, there became a need to distinguish between private and public APIs to offer some stability whilst still developing the software. This culminated in [ADR 060](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-060-go-api-stability.md) which internalized the majority of functionality behind an `internal` directory. This reflected the thinking at the time which endeavoured to present Tendermint as a simple service that one could run without having to worry about what was really happening. For those that had business concerns that required more than what was offered, they had to fork the codebase as is evident with [Celestia](https://github.com/celestiaorg/celestia-core). While there have been numerous discussions over the past years, nothing has been explicitly documented as this RFC aims to do.
```

**Review by thanethomson — APPROVED** [review]
Seems like a reasonable RFC to me. It also seems like it'll still involve quite a bit more discussion before we could make a decision on this front.

**Comment by williambanfield** [comment]
I'm someone opposed to the idea of Tendermint as a library. It's really costly to maintain all of those APIs and we'll end up spending a lot of time on a set of functions that very few users will actually take advantage of. If I had my druthers, ABCI would be close to the entire API of Tendermint and all Apps would likely be out-of-process applications. The RPC overhead isn't very high but the cost of maintaining and testing a separate socket client and internal client is.

**Comment by faddat** [comment]
@williambanfield as a validator I've got to disagree with you:

going from one binary to two is a pretty huge ops burden-- then multiply it by remote signers, then multiply it by 30

so I'd prefer that tendermint generally be compiled into cosmos-sdk chains

(yes, I realize we are discussing beyond the realm of the sdk and into other things, but in general, it will be easier / better for adoption to have one process, as we do now (maybe even better if we think of tm as a library instead of an in-process app)

**Comment by cmwaters** [comment]
> I'm somewhat opposed to the idea of Tendermint as a library. It's really costly to maintain all of those APIs and we'll end up spending a lot of time on a set of functions that very few users will actually take advantage of. If I had my druthers, ABCI would be close to the entire API of Tendermint and all Apps would likely be out-of-process applications. The RPC overhead isn't very high but the cost of maintaining and testing a separate socket client and internal client is.

As we did in v0.35, we can start with a minimal API and expose the components that users constantly request (e.g. the mempool)

**Comment by zmanian** [comment]
Here are some thoughts.

In a world where tendermint was a library, a user you wanted to customize the vote signature system would be able to impotent tendermint via go.mod and share a substantial amount of code with upstream.

the emergence of the modular blockchain architecture where sequencing, data availability, execution and settlement are decoupled has become a prime driver of people writing alt tendermint implementations or build forks. 

if we want to make tendermint a library, our goal should be supporting modular architecture use cases.

**Comment by williambanfield** [comment]
> if we want to make tendermint a library, our goal should be supporting modular architecture use cases.

I would argue that making Tendermint a library and supporting modularity are related, but not necessarily coterminal goals. We can support modularity, in my view, while keeping Tendermint a service (or making it more of one). 

For the example of allowing users to provide their own signature verification schemes, we could very easily imagine a new set of ABCI calls that provide this capability, a `VerifySignature` and `CreateSignature` method that ask the application to verify a vote signature or sign a vote, as well as a new `SignatureMaterial` field added to `PrepareProposal` and `ProcessProposal` arguments that allow the application to provide its own signature material and / or verify the signature provided in the proposal. 

I'll reiterate that this is ultimately my preference. It makes the API boundary clear and prevents the entirety of the Tendermint repo from being covered in our compatibility guarantees.

**Review by williambanfield — APPROVED** [review]

**Comment by cmwaters** [comment]
I think, based from what I'm hearing, a lot of this is more a spectrum and we need to discuss where on that spectrum we think we can provide the most value to the most users. I think part of this discussions needs to be researching how and where do users modify Tendermint. I've personally heard multiple times about having custom mempools. Perhaps it would be valuable for myself and the rest of the team to come up with other cases where a user may want to fork Tendermint i.e. signature system or gossiping protocol.

**Comment by tychoish** [comment]
I think it would be good to find a champion/customer that we could work with (or sponsor?) that was in favor of this work. While I mostly agree with William, if left to my own device. At  the same time it's certianly possible to make tendermint more modular, and if there were users lined up who really wanted this, then we could definitely build that. In order to go down that path, I think it'd be important to do some rough estimation about relative costs and timelines (e.g. tendermint as library could take 2-4x as long, to reach the same level of maturity,) and also (perhaps more importantly,) find some sort of champion user to partner with to build out the correct interfaces and support for their production use cases. Otherwise it sort of feels like we'd be burning a lot of engineering time on "wouldn't it be cool if..."

**Comment by cmwaters** [comment]
Closing in favour of #9459

**Pull request closed without merging**